app/org/sagebionetworks/bridge/dynamodb/DynamoExternalIdDao.java

```diff
@@ -0,0 +1,296 @@
+        DynamoExternalIdentifier identifier = mapper.load(keyObject);
+        if (identifier != null) {
+            identifier.setHealthCode(null);
+            identifier.setReservation(0L);
+            mapper.save(identifier);
+        }
+    }
+    
+    /**
+     * This is intended for testing. Deleting a large number of identifiers will cause DynamoDB capacity exceptions.
+     */
+    @Override
+    public void deleteExternalIds(StudyIdentifier studyId, List<String> externalIds) {
+        checkNotNull(studyId);
+        checkNotNull(externalIds);
+        
+        if (!externalIds.isEmpty()) {
+            List<DynamoExternalIdentifier> idsToDelete = externalIds.stream().map(id -> {
+                return new DynamoExternalIdentifier(studyId, id);
+            }).collect(Collectors.toList());
+            
+            List<FailedBatch> failures = mapper.batchDelete(idsToDelete);
+            BridgeUtils.ifFailuresThrowException(failures);
+        }
+    }
+
+    /**
+     * Get the count query (applies filters) and then sets an offset key and the limit to a page of records, 
+     * plus one, to determine if there are records beyond the current page. 
+     */
+    private DynamoDBQueryExpression<DynamoExternalIdentifier> createGetQuery(StudyIdentifier studyId,
+            String offsetKey, int pageSize, String idFilter, Boolean assignmentFilter) {
+
+        DynamoDBQueryExpression<DynamoExternalIdentifier> query = createCountQuery(studyId, idFilter, assignmentFilter);
+        if (offsetKey != null) {
+            Map<String,AttributeValue> map = new HashMap<>();
+            map.put(STUDY_ID, new AttributeValue().withS(studyId.getIdentifier()));
+            map.put(IDENTIFIER, new AttributeValue().withS(offsetKey));
+            query.withExclusiveStartKey(map);
+        }
+        query.withLimit(pageSize+1);
+        return query;
+    }
+
+    /**
+     * Create a query for records applying the filter values if they exist.
+     */
+    private DynamoDBQueryExpression<DynamoExternalIdentifier> createCountQuery(StudyIdentifier studyId,
+            String idFilter, Boolean assignmentFilter) {
+
+        DynamoDBQueryExpression<DynamoExternalIdentifier> query = new DynamoDBQueryExpression<DynamoExternalIdentifier>();
+        if (idFilter != null) {
+            // You cannot filter a query on a hash key, so we copy this value to another column where we can filter
+            query.withQueryFilterEntry(FILTERABLE_EXTERNAL_ID, new Condition()
+                    .withAttributeValueList(new AttributeValue().withS(idFilter))
+                    .withComparisonOperator(CONTAINS));
+        }
+        if (assignmentFilter == Boolean.TRUE) {
+            query.withQueryFilterEntry(HEALTH_CODE, new Condition().withComparisonOperator(NOT_NULL));
+        } else if (assignmentFilter == Boolean.FALSE) {
+            query.withQueryFilterEntry(HEALTH_CODE, new Condition().withComparisonOperator(NULL));
+        }
```

Comment:
Should the assignmentFilter also account for reserved external IDs? Would that be more useful that assigned external IDs? Would a researcher ever accidentally get a reserved but not assigned ID? Or is this unlikely because of how the API call will include both the reservation and assignment, so this will never come up?